File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/backend/operand/ExStackVarOffset.cpp
// ExStackVarOffset.cpp

#include "ExStackVarOffset.h"

namespace Backend {

ExStackVarOffset::ExStackVarOffset(int val) : AsmOperandImmediate(val) { setType(OperandType::EX_STACK_VAR_OFFSET); }

ExStackVarOffset* ExStackVarOffset::transform(long value) {
    return new ExStackVarOffset(static_cast<int>(value));
}

AsmOperandBasic* ExStackVarOffset::add(int diff) {
    if (getValue() < 0) {
        return new ExStackVarOffset(getValue() + diff);
    } else {
        return this;
    }
}

} // namespace Backend

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/backend/asm/operand/ExStackVarOffset.java
package cn.edu.bit.newnewcc.backend.asm.operand;

/**
 * 栈偏移量，专用于栈帧重新分配的过程.
 */
public class ExStackVarOffset extends Immediate implements ExStackVarAdd {
    ExStackVarOffset(int val) {
        super(val);
    }
    public static ExStackVarOffset transform(long value) {
        return new ExStackVarOffset(Math.toIntExact(value));
    }

    @Override
    public AsmOperand add(int diff) {
        if (getValue() < 0) {
            return new ExStackVarOffset(getValue() + diff);
        } else {
            return this;
        }
    }
}
